src/python/pants/java/nailgun_protocol.py

```diff
@@ -210,14 +215,77 @@ def read_chunk(cls, sock, return_bytes=False):
 
     return chunk_type, payload
 
+  class ProcessStreamTimeout(Exception):
+    """Raised after a timeout completes when a timeout is set on the stream during iteration."""
+
   @classmethod
-  def iter_chunks(cls, sock, return_bytes=False):
-    """Generates chunks from a connected socket until an Exit chunk is sent."""
+  @contextmanager
+  def _set_socket_timeout(cls, sock, timeout=None):
+    """Temporarily set a socket timeout in order to respect a timeout provided to .iter_chunks()."""
+    if timeout is not None:
+      prev_timeout = sock.gettimeout()
+    try:
+      if timeout is not None:
+        sock.settimeout(timeout)
+      yield
+    except socket.timeout:
+      raise cls.ProcessStreamTimeout("socket read timed out with timeout {}".format(timeout))
+    finally:
+      if timeout is not None:
+        sock.settimeout(prev_timeout)
+
+  class TimeoutObject(AbstractClass):
+
+    @abstractmethod
+    def maybe_timeout_options(self):
+      """Called on every stream iteration to obtain a possible specification for a timeout.
+
+      If this method returns non-None, it should return a 2-tuple of the time the timeout began, and
+      the length of the timeout, both as positive floats.
+
+      :returns: a 2-tuple of (timeout_start_time, timeout_interval), or None
```

Comment:
This feels like a great place for a specific `datatype`, as it's a whole interface thing. But happy to leave as-is if you'd prefer :)

Reply:
Agreed!

Reply:
Done!